File: golang/analysis/gotest/builder.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

const baseSpace = "/go/src"

type Builder struct {
	GitCloneURL        string
	GitRef             string
	GtestPackageOrFile string
	GtestParams        string
	ProjectName        string
}

func NewBuilder(envs map[string]string) (*Builder, error) {
	b := &Builder{}
	if envs["GIT_CLONE_URL"] != "" {
		b.GitCloneURL = envs["GIT_CLONE_URL"]
		if b.GitRef = envs["GIT_REF"]; b.GitRef == "" {
			b.GitRef = "master"
		}
	} else if envs["_WORKFLOW_GIT_CLONE_URL"] != "" {
		b.GitCloneURL = envs["_WORKFLOW_GIT_CLONE_URL"]
		b.GitRef = envs["_WORKFLOW_GIT_REF"]
	} else {
		return nil, fmt.Errorf("environment variable GIT_CLONE_URL is required")
	}

	s := strings.TrimSuffix(strings.TrimSuffix(b.GitCloneURL, "/"), ".git")
	b.ProjectName = s[strings.LastIndex(s, "/")+1:]

	//fmt.Println(b.ProjectName)
	b.GtestPackageOrFile = envs["GTEST_PACKAGE_FILE"]
	b.GtestParams = envs["GTEST_PARAMS"]

	return b, nil
}

func (b *Builder) run() error {
	if err := os.Chdir(baseSpace); err != nil {
		return fmt.Errorf("chdir to baseSpace(%s) failed:%v", baseSpace, err)
	}

	if err := b.gitPull(); err != nil {
		return err
	}

	if err := b.gitReset(); err != nil {
		return err
	}

	if err := b.build(); err != nil {
		return err
	}
	return nil
}

func (b *Builder) gitPull() error {
	var command = []string{"git", "clone", "--recurse-submodules", b.GitCloneURL, b.ProjectName}
	if _, err := (CMD{Command: command}).Run(); err != nil {
		fmt.Println("Clone project failed:", err)
		return err
	}
	fmt.Println("Clone project", b.GitCloneURL, "succeded.")
	return nil
}

func (b *Builder) gitReset() error {
	cwd, _ := os.Getwd()
	fmt.Println("current: ", cwd)
	var command = []string{"git", "checkout", b.GitRef, "--"}
	if _, err := (CMD{command, filepath.Join(cwd, b.ProjectName)}).Run(); err != nil {
		fmt.Println("Switch to commit", b.GitRef, "failed:", err)
		return err
	}
	fmt.Println("Switch to", b.GitRef, "succeded.")

	return nil
}

func (b *Builder) build() error {
	cwd, _ := os.Getwd()
	var script string
	if b.GtestPackageOrFile == "" {
		script = fmt.Sprintf("go test %s `go list ./...`", b.GtestParams)
	} else {
		script = fmt.Sprintf("gp test %s %s", b.GtestParams, b.GtestPackageOrFile)
	}
	var command = []string{"sh", "-c", script}
	if _, err := (CMD{command, filepath.Join(cwd, b.ProjectName)}).Run(); err != nil {
		fmt.Printf("Exec: %s failed: %v", script, err)
	}
	fmt.Printf("Exec: %s succeded.\n", script)
	return nil
}

type CMD struct {
	Command []string // cmd with args
	WorkDir string
}

func (c CMD) Run() (string, error) {
	fmt.Println("Run CMD: ", strings.Join(c.Command, " "))

	cmd := exec.Command(c.Command[0], c.Command[1:]...)

	if c.WorkDir != "" {
		cmd.Dir = c.WorkDir
	}

	data, err := cmd.CombinedOutput()
	result := string(data)
	if len(result) > 0 {
		fmt.Println(result)
	}

	return result, err
}

File: python/document/pycco/builder.py
import os
import sys
import json
import subprocess
import pycco.main as pyccoLib
import anymarkup

BASE_SPACE = "/root/src"

class Builder:
    def __init__(self, envs):
        if envs.get('GIT_CLONE_URL'):
            self.git_clone_URL = envs.get('GIT_CLONE_URL').rstrip('/')
            self.git_ref = envs.get('GIT_REF') or 'master'
        elif envs.get('_WORKFLOW_GIT_CLONE_URL'):
            self.git_clone_URL = envs.get('_WORKFLOW_GIT_CLONE_URL').rstrip('/')
            self.git_ref = envs.get('_WORKFLOW_GIT_REF') or 'master'
        else:
            print("environment variable GIT_CLONE_URL is required", flle=sys.stderr)
            exit(1)

        if envs.get('ENTRY_MOD'):
            self.entry_mod = envs.get('ENTRY_MOD')
        else:
            print("environment variable ENTRY_MOD is required", file=sys.stderr)
            exit(1)

        self.project_name = os.path.basename(self.git_clone_URL.rstrip('.git'))
        
    def run(self):
        os.chdir(BASE_SPACE)

        return self.git_pull() and self.git_reset() and self.build() 

    def git_pull(self):
        cmd = ['git', 'clone', '--recurse-submodules', self.git_clone_URL, self.project_name]
        print("Run CMD %s" % ' '.join(cmd), file=sys.stdout)
        r = subprocess.run(cmd)

        if r.returncode != 0:
            print("Git clone failed", file=sys.stderr)
            return False

    def git_reset(self):
        cmd = ['git', 'checkout', self.git_ref, '--']
        print("Run CMD %s" % ' '.join(cmd), file=sys.stdout)
        r = subprocess.run(cmd, cwd=os.path.join(BASE_SPACE, self.project_name), stdout=subprocess.PIPE)
        if r.returncode != 0:
            print("Git checkout failed", file=sys.stderr)
            return False

        return True

    def build(self):
        for root, dirs, files in os.walk(os.path.join(BASE_SPACE, self.project_name)):
            for file_name in files:
                if file_name.endswith('.py'):
                    o = self.pycco(os.path.join(root, file_name))
                    all_true = all_true and o

    def trim_repo_path(self, n):
        return n[len(BASE_SPACE) + 1:]

    def pycco(self, file_name):
        code = open(file_name, 'rb').read().decode('utf-8')
        language = pyccoLib.get_language(file_name, code)
        sections = pyccoLib.parse(code, language)
        pyccoLib.highlight(sections, language, outdir="docs")

        data = { 'file_path': self.trim_repo_path(file_name), 'sections': sections }
        
        print('[COUT] CO_JSON_CONTENT {}'.format(json.dumps(data)))

        return True
